File: src/services/regexEngine.ts
/**
 * ST Regex extension system
 *
 * Regex 扩展兼容系统，支持 global/preset/scoped 三类脚本。
 * 参考: SillyTavern/public/scripts/extensions/regex/engine.js
 */

/**
 * Regex 脚本作用点
 */
export const RegexPlacement = {
  USER_INPUT: 1,
  AI_OUTPUT: 2,
  SLASH_COMMAND: 3,
  WORLD_INFO: 5,
  REASONING: 6,
} as const

export type RegexPlacementValue = (typeof RegexPlacement)[keyof typeof RegexPlacement]

/**
 * 宏替换策略
 */
export const SubstituteRegex = {
  NONE: 0,
  RAW: 1,
  ESCAPED: 2,
} as const

export type SubstituteRegexValue = (typeof SubstituteRegex)[keyof typeof SubstituteRegex]

/**
 * Regex 脚本数据
 */
export interface RegexScriptData {
  id: string
  script_name: string
  find_regex: string
  replace_string: string
  trim_strings?: string[]
  placement?: RegexPlacementValue[]
  disabled?: boolean
  markdown_only?: boolean
  prompt_only?: boolean
  run_on_edit?: boolean
  substitute_regex?: SubstituteRegexValue
  min_depth?: number | null
  max_depth?: number | null
}

/**
 * 创建新的 Regex 脚本
 */
export function createRegexScript(scriptName: string): RegexScriptData {
  return {
    id: crypto.randomUUID(),
    script_name: scriptName,
    find_regex: '',
    replace_string: '',
    trim_strings: [],
    placement: [],
    disabled: false,
    markdown_only: false,
    prompt_only: false,
    run_on_edit: true,
    substitute_regex: SubstituteRegex.NONE,
    min_depth: null,
    max_depth: null,
  }
}

/**
 * Regex Preset 条目
 */
export interface RegexPresetItem {
  id: string
}

/**
 * Regex Preset（启用脚本 ID 列表）
 */
export interface RegexPreset {
  id: string
  name: string
  is_selected?: boolean
  global?: RegexPresetItem[]
  scoped?: RegexPresetItem[]
  preset?: RegexPresetItem[]
}

/**
 * 创建新的 Regex Preset
 */
export function createRegexPreset(name: string): RegexPreset {
  return {
    id: crypto.randomUUID(),
    name,
    is_selected: false,
    global: [],
    scoped: [],
    preset: [],
  }
}

/**
 * Regex 扩展全局设置
 */
export interface RegexExtensionSettings {
  regex: RegexScriptData[]
  regex_presets: RegexPreset[]
  character_allowed_regex: string[]
  preset_allowed_regex: Record<string, string[]>
}

/**
 * 创建默认的 Regex 扩展设置
 */
export function createDefaultRegexExtensionSettings(): RegexExtensionSettings {
  return {
    regex: [],
    regex_presets: [],
    character_allowed_regex: [],
    preset_allowed_regex: {},
  }
}

/**
 * 运行时选项
 */
export interface RegexRunOptions {
  isMarkdown?: boolean
  isPrompt?: boolean
  isEdit?: boolean
  depth?: number | null
  characterOverride?: string
  /** 当前预设 key（用于 preset 脚本授权） */
  presetKey?: string
  /** 当前角色名（用于 scoped 脚本授权） */
  characterName?: string
}

/**
 * 脚本来源类型
 */
export enum ScriptSource {
  Global,
  Preset,
  Scoped,
}

/**
 * 带来源标记的脚本
 */
export interface SourcedScript {
  script: RegexScriptData
  source: ScriptSource
}

/**
 * Regex 执行引擎
 */
export class RegexEngine {
  private regexCache: Map<string, RegExp> = new Map()
  private cacheMaxSize = 1000

  /**
   * 对文本执行 Regex 替换
   *
   * `getRegexedString` 的等价实现。
   * 按照 global -> preset -> scoped 的顺序执行脚本。
   */
  getRegexedString(
    raw: string,
    placement: RegexPlacementValue,
    settings: RegexExtensionSettings,
    options: RegexRunOptions = {}
  ): string {
    // 非字符串输入返回空字符串
    if (typeof raw !== 'string') {
      return ''
    }

    // 空字符串直接返回
    if (raw.length === 0) {
      return raw
    }

    let text = raw

    // 合并允许运行的脚本（按 global -> preset -> scoped 顺序）
    const scripts = this.getAllowedScripts(settings, options)

    // 依次执行每个脚本
    for (const sourcedScript of scripts) {
      if (this.shouldRunScript(sourcedScript.script, placement, options)) {
        text = this.runScript(sourcedScript.script, text, options)
      }
    }

    return text
  }

  /**
   * 获取允许运行的脚本列表（按顺序）
   *
   * 合并顺序：global -> preset -> scoped
   */
  private getAllowedScripts(
    settings: RegexExtensionSettings,
    options: RegexRunOptions
  ): SourcedScript[] {
    const result: SourcedScript[] = []
    const seen = new Set<string>()
    const scriptById = new Map(settings.regex.map(script => [script.id, script]))
    const selectedPresets = settings.regex_presets.filter(preset => preset.is_selected)

    // 1. Global 脚本：没有启用 preset 时保留历史行为，全部 global 可见。
    if (selectedPresets.length === 0) {
      for (const script of settings.regex) {
        if (!seen.has(script.id)) {
          seen.add(script.id)
          result.push({
            script,
            source: ScriptSource.Global,
          })
        }
      }
    } else {
      for (const preset of selectedPresets) {
        for (const item of preset.global ?? []) {
          const script = scriptById.get(item.id)
          if (script && !seen.has(script.id)) {
            seen.add(script.id)
            result.push({ script, source: ScriptSource.Global })
          }
        }
      }
    }

    // 2. Preset 脚本：需要通过 preset_allowed_regex
    if (options.presetKey) {
      const allowedIds = settings.preset_allowed_regex[options.presetKey] ?? []
      for (const preset of selectedPresets) {
        for (const item of preset.preset ?? []) {
          if (!allowedIds.includes(item.id)) continue
          const script = scriptById.get(item.id)
          if (script && !seen.has(script.id)) {
            seen.add(script.id)
            result.push({ script, source: ScriptSource.Preset })
          }
        }
      }
    }

    // 3. Scoped 脚本：需要通过 character_allowed_regex
    const characterScopeEnabled = options.characterName
      ? settings.character_allowed_regex.includes(options.characterName)
      : false
    for (const preset of selectedPresets) {
      for (const item of preset.scoped ?? []) {
        if (!characterScopeEnabled && !settings.character_allowed_regex.includes(item.id)) continue
        const script = scriptById.get(item.id)
        if (script && !seen.has(script.id)) {
          seen.add(script.id)
          result.push({ script, source: ScriptSource.Scoped })
        }
      }
    }

    return result
  }

  /**
   * 判断脚本是否应该运行
   */
  private shouldRunScript(
    script: RegexScriptData,
    placement: RegexPlacementValue,
    options: RegexRunOptions
  ): boolean {
    // 禁用的脚本跳过
    if (script.disabled) {
      return false
    }

    // find_regex 为空跳过
    if (!script.find_regex) {
      return false
    }

    // placement 不包含当前作用点跳过
    if (!script.placement?.includes(placement)) {
      return false
    }

    // 编辑模式下检查 run_on_edit
    if (options.isEdit && !script.run_on_edit) {
      return false
    }

    // 深度过滤
    if (options.depth != null) {
      if (script.min_depth != null && options.depth < script.min_depth) {
        return false
      }
      if (script.max_depth != null && options.depth > script.max_depth) {
        return false
      }
    }

    // markdownOnly 只在 isMarkdown 时运行
    if (script.markdown_only && !options.isMarkdown) {
      return false
    }

    // promptOnly 只在 isPrompt 时运行
    if (script.prompt_only && !options.isPrompt) {
      return false
    }

    // 两者都为 false 时，只在非 markdown、非 prompt 的源文本阶段运行
    if (!script.markdown_only && !script.prompt_only) {
      if (options.isMarkdown || options.isPrompt) {
        return false
      }
    }

    return true
  }

  /**
   * 执行单个脚本
   */
  private runScript(
    script: RegexScriptData,
    text: string,
    _options: RegexRunOptions
  ): string {
    // 编译正则
    const regex = this.compileRegex(script.find_regex, script.substitute_regex)
    if (!regex) {
      return text
    }

    // 执行替换，支持 $1/$<name>，并补充 {{match}}。
    let result = text.replace(regex, (...args: unknown[]) => {
      const match = String(args[0] ?? '')
      const captures = args.slice(1, -2).map(value => String(value ?? ''))
      let replacement = script.replace_string.replaceAll('{{match}}', match)
      replacement = replacement.replace(/\$(\d+)/g, (_, index: string) => {
        const capture = captures[Number(index) - 1]
        return capture ?? ''
      })
      return replacement
    })

    // 处理 trim_strings
    if (script.trim_strings && script.trim_strings.length > 0) {
      result = this.applyTrimStrings(result, script.trim_strings)
    }

    return result
  }

  /**
   * 编译正则表达式
   */
  private compileRegex(pattern: string, _substitute?: SubstituteRegexValue): RegExp | null {
    // 尝试从缓存获取
    if (this.regexCache.has(pattern)) {
      return this.regexCache.get(pattern)!
    }

    // 编译正则
    try {
      // 尝试解析 /pattern/flags 格式
      const { regexPattern, flags } = this.parseRegexPattern(pattern)
      const regex = new RegExp(regexPattern, flags)

      // 管理缓存大小
      if (this.regexCache.size >= this.cacheMaxSize) {
        const firstKey = this.regexCache.keys().next().value
        if (firstKey) {
          this.regexCache.delete(firstKey)
        }
      }
      this.regexCache.set(pattern, regex)

      return regex
    } catch {
      return null
    }
  }

  /**
   * 解析正则表达式模式
   *
   * 支持 /pattern/flags 格式，也支持普通 pattern。
   */
  private parseRegexPattern(pattern: string): { regexPattern: string; flags: string } {
    // 检查是否是 /pattern/flags 格式
    if (pattern.startsWith('/')) {
      const lastSlash = pattern.lastIndexOf('/')
      if (lastSlash > 1) {
        const inner = pattern.slice(1, lastSlash)
        const flagsStr = pattern.slice(lastSlash + 1)
        return { regexPattern: inner, flags: flagsStr }
      }
    }

    // 普通 pattern
    return { regexPattern: pattern, flags: 'g' }
  }

  /**
   * 应用 trim_strings
   */
  private applyTrimStrings(text: string, trimStrings: string[]): string {
    let result = text
    for (const trimStr of trimStrings) {
      // TODO: 执行宏替换
      result = result.replaceAll(trimStr, '')
    }
    return result
  }

  /**
   * 清除缓存
   */
  clearCache(): void {
    this.regexCache.clear()
  }
}

/**
 * 创建默认的 Regex 引擎实例
 */
export function createRegexEngine(): RegexEngine {
  return new RegexEngine()
}

/**
 * 创建默认的运行时选项
 */
export function createDefaultRegexRunOptions(): RegexRunOptions {
  return {
    isMarkdown: false,
    isPrompt: false,
    isEdit: false,
    depth: null,
  }
}

File: src-tauri/src/st/regex_engine.rs
//! ST Regex extension system
//!
//! Regex 扩展兼容系统，支持 global/preset/scoped 三类脚本。
//! 参考: SillyTavern/public/scripts/extensions/regex/engine.js

use regex::Regex;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Regex 脚本作用点
pub struct RegexPlacement;

impl RegexPlacement {
    pub const USER_INPUT: i32 = 1;
    pub const AI_OUTPUT: i32 = 2;
    pub const SLASH_COMMAND: i32 = 3;
    pub const WORLD_INFO: i32 = 5;
    pub const REASONING: i32 = 6;
}

/// 宏替换策略
pub struct SubstituteRegex;

impl SubstituteRegex {
    pub const NONE: i32 = 0;
    pub const RAW: i32 = 1;
    pub const ESCAPED: i32 = 2;
}

/// Regex 脚本数据
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RegexScriptData {
    pub id: String,
    pub script_name: String,
    pub find_regex: String,
    pub replace_string: String,
    #[serde(default)]
    pub trim_strings: Vec<String>,
    #[serde(default)]
    pub placement: Vec<i32>,
    #[serde(default)]
    pub disabled: bool,
    #[serde(default)]
    pub markdown_only: bool,
    #[serde(default)]
    pub prompt_only: bool,
    #[serde(default)]
    pub run_on_edit: bool,
    #[serde(default)]
    pub substitute_regex: i32,
    #[serde(default)]
    pub min_depth: Option<i32>,
    #[serde(default)]
    pub max_depth: Option<i32>,
}

impl RegexScriptData {
    /// 创建新的脚本（带随机 ID）
    pub fn new(script_name: &str) -> Self {
        Self {
            id: uuid::Uuid::new_v4().to_string(),
            script_name: script_name.to_string(),
            find_regex: String::new(),
            replace_string: String::new(),
            trim_strings: Vec::new(),
            placement: Vec::new(),
            disabled: false,
            markdown_only: false,
            prompt_only: false,
            run_on_edit: true,
            substitute_regex: SubstituteRegex::NONE,
            min_depth: None,
            max_depth: None,
        }
    }
}

/// Regex 扩展全局设置
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct RegexExtensionSettings {
    #[serde(default)]
    pub regex: Vec<RegexScriptData>,
    #[serde(default)]
    pub regex_presets: Vec<RegexPreset>,
    #[serde(default)]
    pub character_allowed_regex: Vec<String>,
    #[serde(default)]
    pub preset_allowed_regex: HashMap<String, Vec<String>>,
}

/// Regex Preset（启用脚本 ID 列表）
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RegexPreset {
    pub id: String,
    pub name: String,
    #[serde(default)]
    pub is_selected: bool,
    #[serde(default)]
    pub global: Vec<RegexPresetItem>,
    #[serde(default)]
    pub scoped: Vec<RegexPresetItem>,
    #[serde(default)]
    pub preset: Vec<RegexPresetItem>,
}

impl RegexPreset {
    pub fn new(name: &str) -> Self {
        Self {
            id: uuid::Uuid::new_v4().to_string(),
            name: name.to_string(),
            is_selected: false,
            global: Vec::new(),
            scoped: Vec::new(),
            preset: Vec::new(),
        }
    }
}

/// Regex Preset 条目
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RegexPresetItem {
    pub id: String,
}

/// 运行时选项
#[derive(Debug, Clone, Default)]
pub struct RegexRunOptions {
    pub is_markdown: bool,
    pub is_prompt: bool,
    pub is_edit: bool,
    pub depth: Option<i32>,
    pub character_override: Option<String>,
    /// 当前预设 key（用于 preset 脚本授权）
    pub preset_key: Option<String>,
    /// 当前角色名（用于 scoped 脚本授权）
    pub character_name: Option<String>,
}

/// 脚本来源类型
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ScriptSource {
    Global,
    Preset,
    Scoped,
}

/// 带来源标记的脚本
#[derive(Debug, Clone)]
pub struct SourcedScript {
    pub script: RegexScriptData,
    pub source: ScriptSource,
}

/// Regex 执行引擎
pub struct RegexEngine {
    /// 正则缓存
    regex_cache: lru::LruCache<String, Regex>,
}

impl RegexEngine {
    /// 创建新的引擎
    pub fn new() -> Self {
        Self {
            regex_cache: lru::LruCache::new(std::num::NonZeroUsize::new(1000).unwrap()),
        }
    }

    /// 对文本执行 Regex 替换
    ///
    /// `get_regexed_string` 的等价实现。
    /// 按照 global -> preset -> scoped 的顺序执行脚本。
    pub fn get_regexed_string(
        &mut self,
        raw: &str,
        placement: i32,
        settings: &RegexExtensionSettings,
        options: &RegexRunOptions,
    ) -> String {
        // 非字符串输入返回空字符串
        if raw.is_empty() {
            return String::new();
        }

        // 空字符串直接返回
        let mut text = raw.to_string();

        // 合并允许运行的脚本（按 global -> preset -> scoped 顺序）
        let scripts = self.get_allowed_scripts(settings, options);

        // 依次执行每个脚本
        for sourced_script in scripts {
            if self.should_run_script(&sourced_script.script, placement, options) {
                text = self.run_script(&sourced_script.script, &text, options);
            }
        }

        text
    }

    /// 获取允许运行的脚本列表（按顺序）
    ///
    /// 合并顺序：global -> preset -> scoped
    fn get_allowed_scripts(
        &self,
        settings: &RegexExtensionSettings,
        options: &RegexRunOptions,
    ) -> Vec<SourcedScript> {
        let mut result = Vec::new();

        // 1. Global 脚本：全部可见
        for script in &settings.regex {
            result.push(SourcedScript {
                script: script.clone(),
                source: ScriptSource::Global,
            });
        }

        // 2. Preset 脚本：需要通过 preset_allowed_regex
        // TODO: 从当前预设加载脚本
        // 当前预设名需要在 options.preset_key 中传递
        // if let Some(preset_key) = &options.preset_key {
        //     if let Some(allowed_ids) = settings.preset_allowed_regex.get(preset_key) {
        //         // 只添加在 allow list 中的脚本
        //     }
        // }

        // 3. Scoped 脚本：需要通过 character_allowed_regex
        // TODO: 从当前角色卡加载脚本
        // 当前角色名需要在 options.character_name 中传递
        // if let Some(character_name) = &options.character_name {
        //     if settings.character_allowed_regex.contains(character_name) {
        //         // 添加角色卡内嵌脚本
        //     }
        // }

        result
    }

    /// 判断脚本是否应该运行
    fn should_run_script(
        &self,
        script: &RegexScriptData,
        placement: i32,
        options: &RegexRunOptions,
    ) -> bool {
        // 禁用的脚本跳过
        if script.disabled {
            return false;
        }

        // find_regex 为空或无效跳过
        if script.find_regex.is_empty() {
            return false;
        }

        // placement 不包含当前作用点跳过
        if !script.placement.contains(&placement) {
            return false;
        }

        // 编辑模式下检查 run_on_edit
        if options.is_edit && !script.run_on_edit {
            return false;
        }

        // 深度过滤
        if let Some(depth) = options.depth {
            if let Some(min_depth) = script.min_depth {
                if depth < min_depth {
                    return false;
                }
            }
            if let Some(max_depth) = script.max_depth {
                if depth > max_depth {
                    return false;
                }
            }
        }

        // markdownOnly 只在 is_markdown 时运行
        if script.markdown_only && !options.is_markdown {
            return false;
        }

        // promptOnly 只在 is_prompt 时运行
        if script.prompt_only && !options.is_prompt {
            return false;
        }

        // 两者都为 false 时，只在非 markdown、非 prompt 的源文本阶段运行
        if !script.markdown_only && !script.prompt_only {
            if options.is_markdown || options.is_prompt {
                return false;
            }
        }

        true
    }

    /// 执行单个脚本
    fn run_script(&mut self, script: &RegexScriptData, text: &str, _options: &RegexRunOptions) -> String {
        // 编译正则
        let regex = match self.compile_regex(&script.find_regex, script.substitute_regex) {
            Some(re) => re,
            None => return text.to_string(),
        };

        // 执行替换
        let result = regex.replace_all(text, &script.replace_string);

        // TODO: 处理 trim_strings
        // TODO: 处理宏替换

        result.to_string()
    }

    /// 编译正则表达式
    fn compile_regex(&mut self, pattern: &str, _substitute: i32) -> Option<Regex> {
        // 尝试从缓存获取
        if let Some(re) = self.regex_cache.get(pattern) {
            return Some(re.clone());
        }

        // 编译正则
        match Regex::new(pattern) {
            Ok(re) => {
                self.regex_cache.put(pattern.to_string(), re.clone());
                Some(re)
            }
            Err(_) => None,
        }
    }

    /// 清除缓存
    pub fn clear_cache(&mut self) {
        self.regex_cache.clear();
    }
}

impl Default for RegexEngine {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_basic_replace() {
        let mut engine = RegexEngine::new();
        let script = RegexScriptData {
            id: "test".to_string(),
            script_name: "test".to_string(),
            find_regex: r"hello".to_string(),
            replace_string: "hi".to_string(),
            placement: vec![RegexPlacement::USER_INPUT],
            disabled: false,
            markdown_only: false,
            prompt_only: false,
            run_on_edit: true,
            substitute_regex: SubstituteRegex::NONE,
            min_depth: None,
            max_depth: None,
            trim_strings: Vec::new(),
        };

        let settings = RegexExtensionSettings {
            regex: vec![script],
            ..Default::default()
        };

        let options = RegexRunOptions::default();

        let result = engine.get_regexed_string(
            "hello world",
            RegexPlacement::USER_INPUT,
            &settings,
            &options,
        );

        assert_eq!(result, "hi world");
    }

    #[test]
    fn test_placement_filter() {
        let mut engine = RegexEngine::new();
        let script = RegexScriptData {
            id: "test".to_string(),
            script_name: "test".to_string(),
            find_regex: r"hello".to_string(),
            replace_string: "hi".to_string(),
            placement: vec![RegexPlacement::USER_INPUT],
            disabled: false,
            markdown_only: false,
            prompt_only: false,
            run_on_edit: true,
            substitute_regex: SubstituteRegex::NONE,
            min_depth: None,
            max_depth: None,
            trim_strings: Vec::new(),
        };

        let settings = RegexExtensionSettings {
            regex: vec![script],
            ..Default::default()
        };

        let options = RegexRunOptions::default();

        // AI_OUTPUT 不匹配 USER_INPUT placement
        let result = engine.get_regexed_string(
            "hello world",
            RegexPlacement::AI_OUTPUT,
            &settings,
            &options,
        );

        assert_eq!(result, "hello world");
    }

    #[test]
    fn test_disabled_script() {
        let mut engine = RegexEngine::new();
        let script = RegexScriptData {
            id: "test".to_string(),
            script_name: "test".to_string(),
            find_regex: r"hello".to_string(),
            replace_string: "hi".to_string(),
            placement: vec![RegexPlacement::USER_INPUT],
            disabled: true,
            markdown_only: false,
            prompt_only: false,
            run_on_edit: true,
            substitute_regex: SubstituteRegex::NONE,
            min_depth: None,
            max_depth: None,
            trim_strings: Vec::new(),
        };

        let settings = RegexExtensionSettings {
            regex: vec![script],
            ..Default::default()
        };

        let options = RegexRunOptions::default();

        let result = engine.get_regexed_string(
            "hello world",
            RegexPlacement::USER_INPUT,
            &settings,
            &options,
        );

        assert_eq!(result, "hello world");
    }

    #[test]
    fn test_depth_filter() {
        let mut engine = RegexEngine::new();
        let script = RegexScriptData {
            id: "test".to_string(),
            script_name: "test".to_string(),
            find_regex: r"hello".to_string(),
            replace_string: "hi".to_string(),
            placement: vec![RegexPlacement::USER_INPUT],
            disabled: false,
            markdown_only: false,
            prompt_only: false,
            run_on_edit: true,
            substitute_regex: SubstituteRegex::NONE,
            min_depth: Some(2),
            max_depth: Some(5),
            trim_strings: Vec::new(),
        };

        let settings = RegexExtensionSettings {
            regex: vec![script],
            ..Default::default()
        };

        // depth = 3，在 [2, 5] 范围内
        let options = RegexRunOptions {
            depth: Some(3),
            ..Default::default()
        };

        let result = engine.get_regexed_string(
            "hello world",
            RegexPlacement::USER_INPUT,
            &settings,
            &options,
        );

        assert_eq!(result, "hi world");

        // depth = 1，不在 [2, 5] 范围内
        let options = RegexRunOptions {
            depth: Some(1),
            ..Default::default()
        };

        let result = engine.get_regexed_string(
            "hello world",
            RegexPlacement::USER_INPUT,
            &settings,
            &options,
        );

        assert_eq!(result, "hello world");
    }

    #[test]
    fn test_prompt_only() {
        let mut engine = RegexEngine::new();
        let script = RegexScriptData {
            id: "test".to_string(),
            script_name: "test".to_string(),
            find_regex: r"hello".to_string(),
            replace_string: "hi".to_string(),
            placement: vec![RegexPlacement::USER_INPUT],
            disabled: false,
            markdown_only: false,
            prompt_only: true,
            run_on_edit: true,
            substitute_regex: SubstituteRegex::NONE,
            min_depth: None,
            max_depth: None,
            trim_strings: Vec::new(),
        };

        let settings = RegexExtensionSettings {
            regex: vec![script],
            ..Default::default()
        };

        // is_prompt = true 时运行
        let options = RegexRunOptions {
            is_prompt: true,
            ..Default::default()
        };

        let result = engine.get_regexed_string(
            "hello world",
            RegexPlacement::USER_INPUT,
            &settings,
            &options,
        );

        assert_eq!(result, "hi world");

        // is_prompt = false 时不运行
        let options = RegexRunOptions::default();

        let result = engine.get_regexed_string(
            "hello world",
            RegexPlacement::USER_INPUT,
            &settings,
            &options,
        );

        assert_eq!(result, "hello world");
    }
}
